zilliqa/src/consensus.rs

```diff
@@ -1394,42 +1416,42 @@ impl Consensus {
         }
     }
 
-    fn check_and_commit(&mut self, proposal_hash: Hash) -> Result<()> {
-        let Some(proposal) = self.get_block(&proposal_hash)? else {
-            trace!("block not found: {proposal_hash}");
-            return Ok(());
+    fn check_and_commit(&mut self, proposal: &Block) -> Result<()> {
+        // The condition for a block to be finalized is if there is a direct two-chain. From the paper:
+        // Once a replica is convinced, it checks
+        // if a two-chain is formed over the top of the parent of the
+        // block pointed by the highQC (the first chain in the two-chain
+        // formed has to be a one-direct chain in case of pipelined Fast-
+        // HotStuff). Then a replica can safely commit the parent of the
+        // block pointed by the highQC.
+        // So, in short, look up parent of QC, and finalize it iff the two subsequent blocks
+        // have views N+1, N+2 (the final one being proposal block).
+
+        let Some(qc_block) = self.get_block(&proposal.qc.block_hash)? else {
+            warn!("missing qc block when checking whether to finalize!");
+            return Err(MissingBlockError::from(proposal.qc.block_hash).into());
         };
-        let Some(prev_1) = self.get_block(&proposal.qc.block_hash)? else {
-            trace!("parent not found: {}", proposal.qc.block_hash);
+
+        // At genesis it could be fine not to have a qc block, so don't error.
```

Comment:
Iff genesis is the only reason we don't error then I'd honestly make an explicit check for that, and still error if `qc_block.view() > 0`, only returning Ok(()) if it's actually genesis.